### ISSUE TITLE
[BUG] Forecasts are off when using `Differencer` within `FhPlexForecaster` 

### ISSUE THREAD

**Issue by hliebert** [sktime/sktime#4918]
**Describe the bug**
Forecasts are off when using `Differencer` in a pipeline nested within `FhPlexForecaster`. It seems as if the `inverse_transform` is not applied correctly.

**To Reproduce**
Include any `Differencer` transformation in a forecaster nested within `FhPlexForecaster`. The code below showcases the example (just comment the line applying the `Differencer` in the `TransformedTargetForecaster` pipe). It occurs independent of the how the `na_handling` parameter is chosen.

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction, FhPlexForecaster, ForecastingPipeline, TransformedTargetForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.forecasting.model_selection import ForecastingRandomizedSearchCV
from sktime.performance_metrics.forecasting import MeanSquaredError
from sklearn.preprocessing import MinMaxScaler
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.subset import IndexSubset


y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 4), is_relative=True)

random_forest = RandomForestRegressor()
reduction_forecaster = make_reduction(random_forest, window_length=2, strategy="direct", windows_identical=False)

# Inner pipeline (transformers for X, forecaster)
forecasting_pipeline = ForecastingPipeline(
    [
        ("index_subset", IndexSubset(index_treatment="remove")),
        ("min_max_scaler", TabularToSeriesAdaptor(MinMaxScaler())),
        ("reduction_forecaster", reduction_forecaster),
    ]
)

# Outer pipeline (transformers for y, inner forecasting pipe)
transformed_target_forecaster = TransformedTargetForecaster(
    [
        ("log_transformer", LogTransformer()),
        ("first_differencer", Differencer(lags=1, na_handling="drop_na")),
        ("min_max_scaler", TabularToSeriesAdaptor(MinMaxScaler())),
        ("forecasting_pipeline", forecasting_pipeline),
    ]
)

transformed_target_forecaster.get_params()
grid = {"forecasting_pipeline__reduction_forecaster__estimator__max_features": np.linspace(0.1, 0.9, num=20)}

expanding_window_cv = ExpandingWindowSplitter(fh=horizon, step_length=1)

cv_model = ForecastingRandomizedSearchCV(
    forecaster=transformed_target_forecaster,
    # forecaster=reduction_forecaster,
    cv=expanding_window_cv,
    param_distributions=grid,
    n_iter=5,
    scoring=MeanSquaredError(),
    strategy="refit",
    refit=True,
    return_n_best_forecasters=1,
    update_behaviour="full_refit",
    error_score="raise",
    verbose=1,
)

horizon_specific_forecaster = FhPlexForecaster(forecaster=cv_model)

horizon_specific_forecaster.fit(y, X, fh=horizon)
horizon_specific_forecaster.predict(fh=horizon, X=X)
```

**Expected behavior**
A forecast with the inverse transformations applied correctly.

**Versions**
<details>
System:
    python: 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)]
executable: C:\Program Files\python-3.8.10\python.exe
   machine: Windows-10-10.0.19044-SP0
Python dependencies:
          pip: 21.1.1
       sktime: 0.20.1
      sklearn: 1.2.2
       skbase: 0.4.6
        numpy: 1.23.5
        scipy: 1.10.1
       pandas: 1.5.3
   matplotlib: 3.7.2
       joblib: 1.3.1
  statsmodels: 0.14.0
        numba: 0.57.1
     pmdarima: 2.0.3
      tsfresh: None
   tensorflow: None
tensorflow_probability: None
C:\Program Files\JetBrains\PyCharm 2022.1\plugins\python\helpers\pydev\_pydevd_bundle\pydevd_utils.py:606: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for item in s.iteritems():
</details>

**Comment by fkiraly** [comment]
My suspicion is, this happens whenever the `fh` is non-contiguous with the data seen, and the model/trasformer is one that assumes regular data. E.g., `fh = [2]`. Because then, the differencer will not have a value for `fh = [1]` - it needs this value to invert the differencing.

Interesting API designn question, where do we get it from? A naive way would e to "ask the forecaster", ut composition should be non-contextual, i.e., the differencer should not interact with the forecaster eyond its pulic interfaces.

**Comment by hliebert** [comment]
I can see how this is tricky (and I guess it only makes sense if the user supplies a contiguous forecast horizon in the first place).

**Comment by fkiraly** [comment]
> and I guess it only makes sense if the user supplies a contiguous forecast horizon in the first place

Yes, but that's not a workaround in the `FhPlexForecaster`! Because it always passes non-contiguous `fh` to the inner estimators.

**Comment by fkiraly** [comment]
I think a solution would be a parameter that controls whether internally we pass `[1]`, `[2]`, `[3]` - or `[1]`, `[1, 2]`, `[1, 2, 3]` etc and then subset the output to `[1]`, `[2]`, `[3]`.

**Comment by fkiraly** [comment]
@hliebert, since you mention, you are migrating a forecaster in your internal code base - how is this inversion step done there? What value is used for the index 1, to obtain the inverse difference (cumsum) for index 2?

**Comment by fkiraly** [comment]
You get option 2 (forecaster for index 1) if you have the differencer outside of the `FhPlexTransformer`.

For option 3, we´d need the extra parameter.

**Comment by hliebert** [comment]
@fkiraly All data is currently differenced initially, we are working with rates only and only transform back to levels after the final forecast (akin to option 2). This is tedious in some ways (but considerably more convenient with sktime).  Option 3 sounds nice but also finicky. Maybe just differencing outside is sufficient - I don't have a strong preference for either way.

**Comment by fkiraly** [comment]
Ah, too late, I think I already implemented option 3 here: https://github.com/sktime/sktime/pull/4926

Testing would be appreciated, if you find a minute or two.

**Comment by fkiraly** [comment]
I am also noticing a drawback: in option 3, if the grid search is now inside, it will evaluate on the full `fh`, not just the 3rd item. Which is sth we wanted to avoid, I think.

**Comment by hliebert** [comment]
Thanks, that was quick. I'll test in a bit (can't pull directly from Github at work unfortunately). I'll need to think about the grid search evaluation and how desirable/undesirable evaluation over the full horizon up to `t+N` is.  I've skimmed over the commit and I think you mixed up the labels for the `fh_contiguous` argument in the docstring.

**Comment by fkiraly** [comment]
> I've skimmed over the commit and I think you mixed up the labels for the fh_contiguous argument in the docstring.

Yes, thanks for the review - fixed
